src/variational/advi.jl

```diff
@@ -0,0 +1,105 @@
+"""
+    ADVI(samples_per_step = 10, max_iters = 5000)
+
+Automatic Differentiation Variational Inference (ADVI) for a given model.
+"""
+struct ADVI{AD} <: VariationalInference{AD}
+    samples_per_step # number of samples used to estimate the ELBO in each optimization step
+    max_iters        # maximum number of gradient steps used in optimization
+end
+
+ADVI(args...) = ADVI{ADBackend()}(args...)
+ADVI() = ADVI(10, 5000)
+
+alg_str(::ADVI) = "ADVI"
+
+vi(model::Model, alg::ADVI; optimizer = ADAGrad()) = begin
+    # setup
+    var_info = VarInfo()
+    model(var_info, SampleFromUniform())
+    num_params = size(var_info.vals, 1)
+
+    dists = var_info.dists
+    ranges = var_info.ranges
+    
+    q = MeanField(zeros(num_params), zeros(num_params), dists, ranges)
+
+    # construct objective
+    elbo = ELBO()
+
+    Turing.DEBUG && @debug "Optimizing ADVI..."
+    θ = optimize(elbo, alg, q, model; optimizer = optimizer)
+    μ, ω = θ[1:length(q)], θ[length(q) + 1:end]
+
+    # TODO: make mutable instead?
+    MeanField(μ, ω, dists, ranges) 
+end
+
+# TODO: implement optimize like this?
```

Comment:
Let's stick to the function call version for now.

Reply:
Sounds good :+1: